File: slidingWindow/bestTimeToBuyAndSellStock.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        minPrice = prices[0]
        profit = 0
        for price in prices:
            minPrice = min(minPrice, price)
            profit = max(profit, price - minPrice)
        return profit
    

if __name__ == "__main__":
    prices = [7,1,5,3,6,4]
    print(Solution().maxProfit(prices))

File: slidingWindow/cppRealizations/bestTimeToBuyAndSellStock.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minPrice = prices[0];
        int profit = 0;
        for(int price : prices){
            minPrice = min(minPrice, price);
            profit = max(profit, price - minPrice);
        }
        return profit;
    }
};


int main(){
    vector<int> prices = {7,1,5,3,6,4};
    cout << Solution().maxProfit(prices) << endl;
    return 0;
}